### PR TITLE
0.14.1-beta

### PR DESCRIPTION
**Pull request by blazoncek**
Bugfixes & minor new features

### CODE REVIEW

**Comment by softhack007** [comment]
Hi @blazoncek, thanks for all the work. I'll take a look during this week.

**Comment by blazoncek** [comment]
> Just out of curiosity : How much is many segments ?

For ESP8266, 12 is now already a problem. But of different kind.

**Comment by harueg** [comment]
An only slightly related question:
Is the " shuffle " in a playlist random ?

**Comment by softhack007** [comment]
> An only slightly related question: Is the " shuffle " in a playlist random ?

To avoid drifting too far off-topic... _use the source, Luke_ 😉 

https://github.com/Aircoookie/WLED/blob/1dab26bcbcac051f2b7be47a2d5c757a9938bf1f/wled00/playlist.cpp#L28-L42

**Comment by harueg** [comment]
Thank you !

**Comment by blazoncek** [comment]
@softhack007 apart from cosmetic changes (which are welcome) do you think 0.14.1 could be released as is?

**Comment by softhack007** [comment]
> @softhack007 apart from cosmetic changes (which are welcome) do you think 0.14.1 could be released as is? EDIT: at least as beta?

@blazoncek I think you are right. 0.14.1 has collected many bugfixes. Let's scan over PRs tagged as "0.14.1 candidate", and see what can be merged without danger of breaking other stuff. For example, I'd like to make this small update for pico boards, see my proposal in #3573.

Then we can go for releasing 0.14.1 after the weekend. What do you think? Anything left on your toDo list for 0.14.1?

**Comment by blazoncek** [comment]
I have moved everything, except `nodemcuv2, esp8266_2m, esp01_1m_full, esp32dev, esp32_eth, esp32dev_audioreactive, lolin_s2_mini, esp32c3dev, esp32s3dev_8MB` out of platformio.ini (as an experiment) locally.
It feels so refreshed!

I *do not want* any PR into 0.14.1. Not even @WoodyLetsCode HTML optimisation. Let 0.14.1 be a bugfix release and then concentrate on 0.15 as a way forward.

**Comment by softhack007** [comment]
> It feels so refreshed!

Agreed, so we don't touch platformIO.ini for 0.14.1.

> I _do not want_ any PR into 0.14.1. Not even @WoodyLetsCode HTML optimisation. Let 0.14.1 be a bugfix release and then concentrate on 0.15 as a way forward.

👍 right, let's do it like that.

File: usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      top["Threshold"] = ldrThreshold;
+      top["On Preset"] = ldrOnPreset;
+      top["Off Preset"] = ldrOffPreset;
+  }
+
+  bool readFromConfig(JsonObject& root) {
+      JsonObject top = root[FPSTR(_name)];
+      bool configComplete = !top.isNull();
+      configComplete &= getJsonValue(top["Enabled"], ldrEnabled);
+      configComplete &= getJsonValue(top["LDR Pin"], ldrPin);
```
Comment (softhack007) [inline usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h:91]: This usermod should register `ldrPin` in pinManager, before reading from it.

File: wled00/FX.cpp [hunk 2]
```diff
@@ -604,22 +604,36 @@ static const char _data_FX_MODE_TWINKLE[] PROGMEM = "Twinkle@!,!;!,!;!;;m12=0";
...
+    memset(SEGMENT.data, 0xFF, dataSize); // start by fading pixels up
+    SEGENV.aux0 = 1;
+  }
+
   for (int j = 0; j <= SEGLEN / 15; j++) {
     if (random8() <= SEGMENT.intensity) {
-      for (size_t times = 0; times < 10; times++) //attempt to spawn a new pixel 10 times
-      {
-        uint16_t i = random16(SEGLEN);
+      for (size_t times = 0; times < 10; times++) { //attempt to spawn a new pixel 10 times
```
Comment (softhack007) [inline wled00/FX.cpp:617]: (minor) could use `unsigned` instead of `size_t`

File: wled00/FX.cpp [hunk 3]
```diff
@@ -2590,14 +2605,14 @@ uint16_t mode_twinklefox()
...
-static const char _data_FX_MODE_TWINKLEFOX[] PROGMEM = "Twinklefox@!,Twinkle rate,,,,Cool;;!";
+static const char _data_FX_MODE_TWINKLEFOX[] PROGMEM = "Twinklefox@!,Twinkle rate,,,,Cool;!,!;!";
 
 
 uint16_t mode_twinklecat()
 {
   return twinklefox_base(true);
 }
-static const char _data_FX_MODE_TWINKLECAT[] PROGMEM = "Twinklecat@!,Twinkle rate,,,,Cool;;!";
+static const char _data_FX_MODE_TWINKLECAT[] PROGMEM = "Twinklecat@!,Twinkle rate,,,,Cool;!,!;!";
```
Comment (softhack007) [inline wled00/FX.cpp:2615]: side note: I think we should add some documentation for effect metadata format into the KB. 
I'm still struggling to understand when to use `,` `;` `!` 😅

File: wled00/FX.h [hunk 4]
```diff
@@ -109,20 +109,15 @@
...
-// bit    0: segment is selected
+#define DARKSLATEGREY DARKSLATEGRAY
+
+// segment options
 #define NO_OPTIONS   (uint16_t)0x0000
-#define TRANSPOSED   (uint16_t)0x0400 // rotated 90deg & reversed
-#define REVERSE_Y_2D (uint16_t)0x0200
-#define MIRROR_Y_2D  (uint16_t)0x0100
-#define TRANSITIONAL (uint16_t)0x0080
+#define TRANSPOSED   (uint16_t)0x0100 // rotated 90deg & reversed
```
Comment (softhack007) [inline wled00/FX.h:116]: just curious - is this change of constants "transparent", i.e. does not impact presets from 0.14.0 ?

File: usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    static const char _name[];
+
+  public:
+    void setup() {
+    }
+
+    void loop() {
+      // Only update every 10 seconds
+      if (millis() - lastMillis > 10000) {      
+        if (ldrEnabled == true) {
```
Comment (softhack007) [inline usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h:30]: loop should do nothing if not enabled: `if (!ldrEnabled) return;`

File: usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+          // Default state is off
+          if (ldrEnabledPreviously == false) {
+              applyPreset(ldrOffPreset);
+              ldrEnabledPreviously = true;
+              ldrLEDState = 0;
+          }
+
+          // Get LDR reading and increment counter by number of seconds since last read
+          ldrReading = analogRead(ldrPin);
```
Comment (softhack007) [inline usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h:39]: newer frameworks will hate you (and throw thousands of waning) if trying to read from an invalid (-1 or not ADC) pin.

Suggestion: add this
```C++
        if ((ldrPin >= 0) && (digitalPinToAnalogChannel(ldrPin) >= 0)) // avoid reading from invalid pin
            ldrReading = analogRead(ldrPin);
```

**Review by softhack007 — CHANGES_REQUESTED** [review]
Hi, i've managed to do a fast walk through the changes.
Only very minor comments and a few questions. The only part that should get corrected is the new LDR usermod - it uses pins for analogread, but does not register them in pinmanager. Also it seems to not disable itself when set to "disable".

For everything else, I think we are ready to go 👍

File: usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h (hunk 1 above)
Comment (blazoncek) [reply usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h:91]: Agreed but usermods are entirely user's responsibility and custom compiled (unless we decide to include them into standard builds).
I would recommend to leave it and notify usermod author to make it compliant.

File: wled00/FX.cpp (hunk 3 above)
Comment (blazoncek) [reply wled00/FX.cpp:2615]: It already exists thanks to @scottrbailey : https://kno.wled.ge/interfaces/json-api/#effect-metadata

File: wled00/FX.h (hunk 4 above)
Comment (blazoncek) [reply wled00/FX.h:116]: Yes. It is internal only.

File: wled00/FX.cpp (hunk 2 above)
Comment (blazoncek) [reply wled00/FX.cpp:617]: technically the same, so yes.

File: usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h (hunk 5 above)
Comment (blazoncek) [reply usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h:30]: usermod, so ... 😄

File: usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h (hunk 1 above)
Comment (softhack007) [reply usermods/LDR_Dusk_Dawn_v2/usermod_LDR_Dusk_Dawn_v2.h:91]: Its correct we should leave usermods to their authors ... 
just this time i'll add a quick fix - might safe us a few support questions and maybe prevent crashes.

**Review by softhack007 — APPROVED** [review]

**Comment by blazoncek** [comment]
@Aircoookie please get back to us. 😉 

@softhack007 if @Aircoookie does not respond in a couple of days I think we can merge this and release beta/GM version on Monday/Tuesday and if all goes well release 0.14.1 on Saturday/Sunday just before Christmas. I think it is safe as there were mostly bugfixes included.

**Comment by blazoncek** [comment]
@softhack007 I am merging this now. Can you prepare a release?

**Comment by softhack007** [comment]
@blazoncek I can make the gh "golden master" pre-release tomorrow - just need to come up with a few intro words - as you already put all details into changeling. Tomorrow, because my brains want to sleep after a lo g day now ;-)
Cheers 👋

**Comment by blazoncek** [comment]
I've already saved release draft. It only requires binaries and publish button.

**Comment by softhack007** [comment]
👍 I'll quickly cross read and then push it tomorrow morning g

**Pull request merged**